pallets/unified-accounts/src/impls.rs

```diff
@@ -0,0 +1,175 @@
+    }
+}
+
+/// AddressMapping wrapper implementation
+impl<T: Config> AddressMapping<T::AccountId> for Pallet<T> {
+    fn into_account_id(evm_address: H160) -> T::AccountId {
+        <Self as UnifiedAddressMapper<T::AccountId>>::to_account_id_or_default(&evm_address)
+    }
+}
+
+/// OnKilledAccout hooks implementation for removing storage mapping
+/// for killed accounts
+pub struct KillAccountMapping<T>(PhantomData<T>);
+impl<T: Config> OnKilledAccount<T::AccountId> for KillAccountMapping<T> {
+    fn on_killed_account(who: &T::AccountId) {
+        // remove mapping created by `claim_account` or `get_or_create_evm_address`
+        if let Some(evm_addr) = EvmToNative::<T>::take(who) {
+            NativeToEvm::<T>::remove(evm_addr);
+            EvmToNative::<T>::remove(who);
+        }
+    }
+}
+
+/// A lookup implementation returning the `AccountId` from `MultiAddress::Address20` (EVM Address).
+impl<T: Config> StaticLookup for Pallet<T> {
+    type Source = MultiAddress<T::AccountId, ()>;
+    type Target = T::AccountId;
+
+    fn lookup(a: Self::Source) -> Result<Self::Target, LookupError> {
+        match a {
+            MultiAddress::Address20(i) => Ok(
+                <Self as UnifiedAddressMapper<T::AccountId>>::to_account_id_or_default(
+                    &EvmAddress::from_slice(&i),
+                ),
+            ),
+            _ => Err(LookupError),
+        }
+    }
+
+    fn unlookup(a: Self::Target) -> Self::Source {
+        MultiAddress::Id(a)
+    }
+}
+
+/// EIP-712 compatible signature scheme for verifying ownership of EVM Address
+/// https://eips.ethereum.org/EIPS/eip-712
+///
+/// Raw Data = Domain Separator + Type Hash + keccak256(AccountId)
+pub struct EIP712Signature<T: Config, ChainId: Get<u64>>(PhantomData<(T, ChainId)>);
+impl<T: Config, ChainId: Get<u64>> SignatureHelper for EIP712Signature<T, ChainId> {
+    type AccountId = T::AccountId;
+    /// evm address type
+    type Address = EvmAddress;
+    /// A signature (a 512-bit value, plus 8 bits for recovery ID).
+    type Signature = [u8; 65];
+
+    fn build_signing_payload(who: &Self::AccountId) -> [u8; 32] {
+        let domain_separator = Self::build_domain_separator();
+        let args_hash = Self::build_args_hash(who);
+
+        let mut payload = b"\x19\x01".to_vec();
+        payload.extend_from_slice(&domain_separator);
+        payload.extend_from_slice(&args_hash);
+        keccak_256(&payload)
+    }
+
+    fn verify_signature(who: &Self::AccountId, sig: &Self::Signature) -> Option<EvmAddress> {
+        let payload_hash = Self::build_signing_payload(who);
+
+        sp_io::crypto::secp256k1_ecdsa_recover(sig, &payload_hash)
+            .map(|pubkey| H160::from(H256::from_slice(&keccak_256(&pubkey))))
+            .ok()
+    }
+}
+
+impl<T: Config, ChainId: Get<u64>> EIP712Signature<T, ChainId> {
+    /// TODO: minor, use hardcoded bytes, configurable via generics
+    fn build_domain_separator() -> [u8; 32] {
+        let mut hash =
```

Comment:
```suggestion
        let mut domain =
```

It's domain, not a hash.